11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+				0u,
+				IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER,
+				IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+				IShader::E_SHADER_STAGE::ESS_COMPUTE,
+				1u,
+				&sampler
+			};
+
+			imageFirstAxisFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+		}
+
+		smart_refctd_ptr<IGPUPipelineLayout> lastAxisFFT_convolution_lastAxisIFFTPipelineLayout;
+		{
+			auto sampler = createSampler(ISampler::E_TEXTURE_CLAMP::ETC_MIRROR);
+			smart_refctd_ptr<IGPUSampler> samplers[CHANNELS];
+			std::fill_n(samplers, CHANNELS, sampler);
+			IGPUDescriptorSetLayout::SBinding bnd =
+			{
+				IDescriptorSetLayoutBase::SBindingBase(),
+				0u,
+				IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER,
+				IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+				IShader::E_SHADER_STAGE::ESS_COMPUTE,
+				CHANNELS,
+				samplers
+			};
+
+			lastAxisFFT_convolution_lastAxisIFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+		}
+
+		smart_refctd_ptr<IGPUPipelineLayout> imageFirstAxisIFFTPipelineLayout;
+		{
+			IGPUDescriptorSetLayout::SBinding bnd =
+			{
+				IDescriptorSetLayoutBase::SBindingBase(),
+				0u,
+				IDescriptor::E_TYPE::ET_STORAGE_IMAGE,
+				IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+				IShader::E_SHADER_STAGE::ESS_COMPUTE,
+				1,
+				nullptr
+			};
+
+			imageFirstAxisIFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+		}
+
+		m_readCache = nullptr;
+		m_writeCache = core::make_smart_refctd_ptr<IShaderCompiler::CCache>();
+		auto shaderCachePath = localOutputCWD / "cache.bin";
+
+		{
+			core::smart_refctd_ptr<system::IFile> shaderReadCacheFile;
+			{
+				system::ISystem::future_t<core::smart_refctd_ptr<system::IFile>> future;
+				m_system->createFile(future, shaderCachePath.c_str(), system::IFile::ECF_READ);
+				if (future.wait())
+				{
+					future.acquire().move_into(shaderReadCacheFile);
+					if (shaderReadCacheFile)
+					{
+						const size_t size = shaderReadCacheFile->getSize();
+						if (size > 0ull)
+						{
+							std::vector<uint8_t> contents(size);
+							system::IFile::success_t succ;
+							shaderReadCacheFile->read(succ, contents.data(), 0, size);
+							if (succ)
+								m_readCache = IShaderCompiler::CCache::deserialize(contents);
+						}
+					}
+				}
+				else
+					m_logger->log("Failed Opening Shader Cache File.", ILogger::ELL_ERROR);
+			}
+
+		}
+		
+
+		// Kernel second axis FFT has no descriptor sets so we just create another pipeline with the same layout
+		// TODO: To avoid duplicated layouts we could make samplers dynamic in the first axis FFT. Also if we don't hardcode (by #defining) some stuff in the first axis FFT
+		//		 (once FFT ext is back) we can also avoid having duplicated pipelines (like the old Bloom example, which had a single pipeline for forward FFT along an axis)
+		//       and setting stuff via shader push constants (such as which axis to perform FFT on and the size of output image).
```

Comment:
slap everything in one descriptor set, one pipeline layout, one set of push constants, the less code to read, the better